lib/chewy/type/adapter/active_record.rb

```diff
@@ -22,31 +22,65 @@ def cleanup_default_scope!
           @default_scope = @default_scope.reorder(nil).limit(nil).offset(nil)
         end
 
-        def import_scope(scope, batch_size)
-          scope = scope.reorder(target_id.asc).limit(batch_size)
+        def import_scope(scope, batch_size, sort_by_updated_at, &block)
+          if sort_by_updated_at && timestamp_column_name
+            import_scope_ordered_by_timestamp(scope, batch_size, &block)
+          else
+            import_scope_ordered_by_id(scope, batch_size, &block)
+          end
+        end
 
-          ids = pluck_ids(scope)
+        def import_scope_ordered_by_id(scope, batch_size)
           result = true
 
+          scope = scope.reorder(id_column.asc).limit(batch_size)
+
+          ids = pluck_ids(scope)
+
           while ids.present?
             result &= yield grouped_objects(default_scope_where_ids_in(ids))
             break if ids.size < batch_size
-            ids = pluck_ids(scope.where(target_id.gt(ids.last)))
+            ids = pluck_ids(scope.where(id_column.gt(ids.last)))
           end
 
           result
         end
 
-        def target_id
+        def import_scope_ordered_by_timestamp(scope, batch_size)
+          result = true
+
+          scope = scope.reorder(timestamp_column.asc, id_column.asc).limit(batch_size)
+
+          ids_and_dates = pluck_ids_and_dates(scope)
+
+          while ids_and_dates.present?
+            result &= yield grouped_objects(scope_where_ids_in(scope, ids_and_dates.map(&:first)))
+            break if ids_and_dates.size < batch_size
+            last_id, last_updated_at = ids_and_dates.last
+            ids_and_dates = pluck_ids_and_dates(scope.where(next_timestamp_ordered_batch_condition(last_id, last_updated_at)))
+          end
+
+          result
+        end
+
+        def id_column
           target.arel_table[target.primary_key]
         end
 
+        def timestamp_column
+          target.arel_table[timestamp_column_name]
+        end
+
         def pluck_ids(scope)
           scope.except(:includes).uniq.pluck(target.primary_key.to_sym)
         end
 
+        def pluck_ids_and_dates(scope)
+          scope.except(:includes).uniq.pluck(target.primary_key.to_sym, timestamp_column_name)
```

Comment:
There are 3 solutions that I can see now:

* Use `select(:id, :updated_at)`. That's bad performance-wise.
* Include the backport gem (`pluck_all` was the name, IIRC). Again, an extra dependency is not perfect.
* Roll out our own `pluck_all` by copying the code from somewhere and throwing out stuff that we don't need. Feels dirty.
* There is a way to return array of tules (or hashes) from the database. I've done it a couple of times. Don't remember the details, but it wasn't too hard. `AR::Base.select_all(scope)`, or something like this.

And I guess we need to check whether method `pluck` arity can be > 1 and fall back to our workaround only in case if it is not supported.